src/visions/utils/pandas.py

```diff
@@ -0,0 +1,130 @@
+import attr
+from importlib import util as import_util
+from typing import Callable, Type, List
+import pandas as pd
+
+
+def has_import(module: str) -> bool:
+    has_module = import_util.find_spec(module) is not None
+    return has_module
+
+
+@attr.s
+class Engine:
+    name = attr.ib()
+
+    @classmethod
+    def setup(cls, *args, **kwargs) -> None:
+        raise NotImplementedError("No setup defined for generic engine")
+
+    @staticmethod
+    def apply(series: pd.Series) -> Callable[[Callable], pd.Series]:
+        raise NotImplementedError("No apply defined for generic engine")
+
+
+class PandasEngine(Engine):
+    name = "pandas"
+    _is_setup = False
+
+    @classmethod
+    def setup(cls, *args, **kwargs) -> None:
+        pass
+
+    @staticmethod
+    def apply(series: pd.Series) -> Callable[[Callable], pd.Series]:
+        return series.apply
+
+
+class SwifterEngine(Engine):
+    name = "swifter"
+    _is_setup = False
+
+    @classmethod
+    def setup(cls, *args, **kwargs) -> None:
+        if cls._is_setup:
+            return
+
+        import swifter
+
+        cls._is_setup = True
+
+    @staticmethod
+    def apply(series: pd.Series) -> Callable[[Callable], pd.Series]:
+        return series.swifter.apply
+
+
+class PandarallelEngine(Engine):
+    name = "pandarallel"
+    _is_setup = False
+
+    @classmethod
+    def setup(cls, *args, **kwargs) -> None:
+        if cls._is_setup:
+            return
+
+        from pandarallel import pandarallel
+
+        pandarallel.initialize(*args)
+        cls._is_setup = True
+
+    @staticmethod
+    def apply(series: pd.Series) -> Callable[[Callable], pd.Series]:
+        return series.parallel_apply
+
+
+_PANDAS_ENGINES = [PandasEngine, SwifterEngine, PandarallelEngine]
+
+
+class PandasEnginesCollection:
```

Comment:
Do we really need the PandasEnginesCollection abstraction? I'd say it's cleaner without.

Reply:
Yes we do - not every engine supports every API. For example, pandarallel supports map but swifter only supports apply.